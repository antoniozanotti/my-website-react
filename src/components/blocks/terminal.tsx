import React from "react";
import { useState, useEffect, useRef } from "react";

export interface TerminalProps extends React.ComponentProps<"p"> {}

export default function Terminal({ ...props }: TerminalProps) {
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const lines = [
    "Hello, I'm a Frontend Developer",
    "I was a Full Stack Developer",
    "Are you still here?",
    "Scroll the page, my friend!"
  ];
  let line = useRef(lines[0]);
  let lineIndex = useRef(0);
  let task = useRef("write");
  let timePause = useRef(0);
  const [title, setTitle] = useState("");
  const [classVisibilityCursor, setClassVisibilityCursor] = useState("");

  useEffect(() => {
    const delayPause = 14;
    const interval = setInterval(() => {
      if (
        task.current == "pause-so-erase" ||
        task.current == "pause-so-write"
      ) {
        if (![0,1,2,3,4,10,11,12,13,14].includes(timePause.current)) {
          setClassVisibilityCursor("");
        } else {
          setClassVisibilityCursor("invisible");
        }
        if (timePause.current < delayPause) {
          timePause.current = timePause.current + 1;
        } else {
          setClassVisibilityCursor("");
          timePause.current = 0;
          if (task.current == "pause-so-erase") {
            task.current = "erase";
          } else {
            if (lineIndex.current == lines.length - 1) {
              lineIndex.current = 0;
            } else {
              lineIndex.current = lineIndex.current + 1;
            }
            line.current = lines[lineIndex.current];
            task.current = "write";
          }
        }
      } else if (task.current == "write") {
        if (title.length == line.current.length) {
          task.current = "pause-so-erase";
        } else {
          setTitle(line.current.substring(0, title.length + 1));
        }
      } else {
        if (title.length == 0) {
          task.current = "pause-so-write";
        } else {
          setTitle(line.current.substring(0, title.length - 1));
        }
      }
    }, 60);
    return () => clearInterval(interval);
  }, [title, classVisibilityCursor]);

  let pClasses = "text-c4 sm:text-c3 lg:text-c5 flex items-center leading-3 lg:leading-5";
  props.className = props.className
    ? pClasses + " " + props.className
    : pClasses;
  
  return (
    <p {...props}>
      {title}
      <span
        className={`w-2 lg:w-2.5 h-4 sm:h-4 lg:h-5 bg-primary dark:bg-dark-primary inline-block ml-1 ${classVisibilityCursor}`}
      ></span>
    </p>
  );
}
