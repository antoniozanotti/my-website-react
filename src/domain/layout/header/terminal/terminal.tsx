"use client";

import { useState, useEffect, useRef } from "react";
import {useTranslations} from 'next-intl';

export function Terminal({ ...props }: React.ComponentProps<"p">) {
  const t = useTranslations('Header.terminal');
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const lines = [
    t('line1'),
    t('line2'),
    t('line3'),
    t('line4')
  ];
  let line = useRef(lines[0]);
  let lineIndex = useRef(0);
  let task = useRef("write");
  let timePause = useRef(0);
  const [title, setTitle] = useState("");
  const [cursorIsVisible, setCursorIsVisible] = useState(true);

  useEffect(() => {
    const delayPause = 14;
    const interval = setInterval(() => {
      if (
        task.current == "pause-so-erase" ||
        task.current == "pause-so-write"
      ) {
        if (![0,1,2,3,4,10,11,12,13,14].includes(timePause.current)) {
          setCursorIsVisible(true);
        } else {
          setCursorIsVisible(false);
        }
        if (timePause.current < delayPause) {
          timePause.current = timePause.current + 1;
        } else {
          setCursorIsVisible(true);
          timePause.current = 0;
          if (task.current == "pause-so-erase") {
            task.current = "erase";
          } else {
            if (lineIndex.current == lines.length - 1) {
              lineIndex.current = 0;
            } else {
              lineIndex.current = lineIndex.current + 1;
            }
            line.current = lines[lineIndex.current];
            task.current = "write";
          }
        }
      } else if (task.current == "write") {
        if (title.length == line.current.length) {
          task.current = "pause-so-erase";
        } else {
          setTitle(line.current.substring(0, title.length + 1));
        }
      } else {
        if (title.length == 0) {
          task.current = "pause-so-write";
        } else {
          setTitle(line.current.substring(0, title.length - 1));
        }
      }
    }, 60);
    return () => clearInterval(interval);
  }, [title, cursorIsVisible]);

  const pBaseStyles = "text-c4 sm:text-c3 lg:text-c5 flex items-center leading-3 lg:leading-5";
  props.className = props.className
    ? `${pBaseStyles} ${props.className}`
    : pBaseStyles;
  const cursorBaseStyles = "w-2 lg:w-2.5 h-4 sm:h-4 lg:h-5 bg-primary dark:bg-dark-primary inline-block ml-1";
  const cursorVisibilityStyles = cursorIsVisible ? "" : "invisible";
  
  return (
    <p {...props}>
      {title}
      <span
        className={`${cursorBaseStyles} ${cursorVisibilityStyles}`}
      ></span>
    </p>
  );
}
