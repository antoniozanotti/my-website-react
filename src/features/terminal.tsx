import React, { useState, useEffect, useRef } from "react";

export interface TerminalProps {
  className?: string;
}

export default function Terminal({ className = "" }: TerminalProps) {
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const lines = [
    "I'm a Frontend Developer",
    "I'm a Full Stack Developer",
    "Are you still here? Scroll the page, my friend!",
    "I'm a Software Engineer",
    "I'm a Programmer",
    "Why are there so many positions?",
    "I'm a Web Developer",
    "I'm a Jamstack Developer",
    "I'm aâ€¦ Does it matter?",
  ];
  let line = useRef(lines[0]);
  let lineIndex = useRef(0);
  let task = useRef("write");
  let timePause = useRef(0);
  const [title, setTitle] = useState("");
  const [classVisibilityCursor, setClassVisibilityCursor] = useState("");

  useEffect(() => {
    const delayPause = 20;
    const interval = setInterval(() => {
      if (
        task.current == "pause-so-erase" ||
        task.current == "pause-so-write"
      ) {
        if (![0, 1, 2, 8, 9, 10, 16, 17, 18].includes(timePause.current)) {
          setClassVisibilityCursor("");
        } else {
          setClassVisibilityCursor("invisible");
        }
        if (timePause.current < delayPause) {
          timePause.current = timePause.current + 1;
        } else {
          setClassVisibilityCursor("");
          timePause.current = 0;
          if (task.current == "pause-so-erase") {
            task.current = "erase";
          } else {
            if (lineIndex.current == lines.length - 1) {
              lineIndex.current = 0;
            } else {
              lineIndex.current = lineIndex.current + 1;
            }
            line.current = lines[lineIndex.current];
            task.current = "write";
          }
        }
      } else if (task.current == "write") {
        if (title.length == line.current.length) {
          task.current = "pause-so-erase";
        } else {
          setTitle(line.current.substring(0, title.length + 1));
        }
      } else {
        if (title.length == 0) {
          task.current = "pause-so-write";
        } else {
          setTitle(line.current.substring(0, title.length - 1));
        }
      }
    }, 100);
    return () => clearInterval(interval);
  }, [title, classVisibilityCursor]);
  
  return (
    <p
      className={`text-c2 sm:text-c3 lg:text-c5 flex items-center leading-3 lg:leading-5 ${className}`}
    >
      {title}
      <span
        className={`w-1.5 sm:w-2 lg:w-2.5 h-3 sm:h-4 lg:h-5 bg-primary dark:bg-dark-primary inline-block ml-1 ${classVisibilityCursor}`}
      ></span>
    </p>
  );
}
