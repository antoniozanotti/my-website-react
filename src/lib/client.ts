import { ApolloClient, HttpLink, InMemoryCache } from "@apollo/client";
import { registerApolloClient } from "@apollo/experimental-nextjs-app-support/rsc";

const TOKEN = process.env.CONTENTFUL_ACCESS_TOKEN;
const SPACE = process.env.CONTENTFUL_SPACE_ID;
const ENVIRONMENT = process.env.CONTENTFUL_ENVIRONMENT_ID;
const URL = `https://graphql.contentful.com/content/v1/spaces/${SPACE}/environments/${ENVIRONMENT}`;

export const { getClient } = registerApolloClient(() => {
  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new HttpLink({
      uri: URL,
      headers: {
        Authorization: `Bearer ${TOKEN}`,
      },
    }),
  });
});
