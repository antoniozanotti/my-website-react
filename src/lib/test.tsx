import React, { ReactElement } from "react";
import {
  render,
  RenderOptions,
  RenderResult,
  Screen,
} from "@testing-library/react";
import { Theme } from "@/domain/theme/theme/theme";
import { Providers } from "../app/providers";
import messages from "@/messages/en.json";

function AllTheProviders({ children }: { children: React.ReactNode }) {
  return (
    <Providers locale="en" messages={messages}>
      <Theme>
        <div data-testid="container">{children}</div>
      </Theme>
    </Providers>
  );
}

function customRender(
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
): RenderResult {
  return render(ui, { wrapper: AllTheProviders, ...options });
}

function firstComponent(screen: Screen) {
  return screen.getByTestId("container").firstChild;
}

export * from "@testing-library/react";
export * from "@testing-library/user-event";
export { customRender as render, firstComponent };
import "@testing-library/jest-dom";

jest.mock("next/navigation", () => {
  return {
    useRouter() {
      return { push: () => {} };
    },
    useParams: () => ({}),
  };
});
