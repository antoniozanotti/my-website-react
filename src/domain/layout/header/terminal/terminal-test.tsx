import { test, expect } from '@playwright/experimental-ct-react';
import { Terminal } from "./terminal";

test.describe("Terminal", () => {
  test("should render section", async ({ mount }) => {
    const component = await mount(<Terminal />);
    await expect(component.locator(":scope:is(p)")).toBeVisible();
  });

  test("should accept className", async ({ mount }) => {
    const component = await mount(<Terminal className='test' />);
    await expect(component).toHaveClass(/test$/);
  });
});
