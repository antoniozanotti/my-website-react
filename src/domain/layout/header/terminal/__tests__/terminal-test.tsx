import { screen, render, firstComponent, act } from "@/lib/test";
import { Terminal } from "../terminal";

const seconds = 1000;

describe("Terminal", () => {
  it("should render section", async () => {
    const { container } = render(<Terminal />);
    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 4 * seconds));
    });
    expect(screen.getByRole("paragraph")).toBeVisible();
    expect(container).toMatchSnapshot();
  });

  it("should accept className", async () => {
    const { container } = render(<Terminal className='test' />);
    const component = firstComponent(screen);

    expect(component).toHaveClass(/test$/);
    expect(container).toMatchSnapshot();
  });
});
