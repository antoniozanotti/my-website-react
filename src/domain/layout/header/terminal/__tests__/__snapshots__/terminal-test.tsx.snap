// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Terminal should accept className 1`] = `
<div>
  <div
    class="dark"
  >
    <div
      class="className bg-dark-2 dark:bg-dark-dark-2 text-primary dark:text-dark-primary text-c3 lg:text-c4"
    >
      <div
        data-testid="container"
      >
        <p
          class="text-c4 sm:text-c3 lg:text-c5 flex items-center leading-3 lg:leading-5 test"
        >
          <span
            class="w-2 lg:w-2.5 h-4 sm:h-4 lg:h-5 bg-primary dark:bg-dark-primary inline-block ml-1 "
          />
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Terminal should render section 1`] = `
<div>
  <div
    class="dark"
  >
    <div
      class="className bg-dark-2 dark:bg-dark-dark-2 text-primary dark:text-dark-primary text-c3 lg:text-c4"
    >
      <div
        data-testid="container"
      >
        <p
          class="text-c4 sm:text-c3 lg:text-c5 flex items-center leading-3 lg:leading-5"
        >
          H
          <span
            class="w-2 lg:w-2.5 h-4 sm:h-4 lg:h-5 bg-primary dark:bg-dark-primary inline-block ml-1 "
          />
        </p>
      </div>
    </div>
  </div>
</div>
`;
